### PR TITLE
openstackclient: init at 1.5

### PR DESCRIPTION
**Pull request by emilytrau**
###### Motivation for this change

This brings back the openstack CLI client. It was removed along with the rest of openstack back in #32637. This time it's just the clients, so it should be less of a maintenance hassle. This PR creates all the necessary python packages required in the minimal dependency tree of `openstackclient`. I've stopped at including all of the ([many](https://docs.openstack.org/newton/user-guide/common/cli-overview.html#individual-command-line-clients)) optional plugins to keep the PR size somewhat reasonable, to be added later.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/stestr/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  '';
+  disabledTests = [
+    # Broken test
+    "test_empty_with_pretty_out"
+  ];
+  pythonImportsCheck = [ "stestr" ];
+
+  meta = with lib; {
+    description = "A parallel Python test runner built around subunit";
+    homepage = "http://stestr.readthedocs.io/en/latest/";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/stestr/default.nix:38]:
```suggestion
    homepage = "https://stestr.readthedocs.io/en/latest/";
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 134282` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
35 packages built:
</summary>
<ul>
<li>openstackclient</li>
<li>python38Packages.cinderclient</li>
<li>python38Packages.debtcollector</li>
<li>python38Packages.hacking</li>
<li>python38Packages.keystoneauth1</li>
<li>python38Packages.keystoneclient</li>
<li>python38Packages.novaclient</li>
<li>python38Packages.openstacksdk</li>
<li>python38Packages.os-service-types</li>
<li>python38Packages.osc-lib</li>
<li>python38Packages.oslo-config</li>
<li>python38Packages.oslo-context</li>
<li>python38Packages.oslo-i18n</li>
<li>python38Packages.oslo-log</li>
<li>python38Packages.oslo-serialization</li>
<li>python38Packages.oslo-utils</li>
<li>python38Packages.oslotest</li>
<li>python38Packages.stestr</li>
<li>python39Packages.cinderclient</li>
<li>python39Packages.debtcollector</li>
<li>python39Packages.hacking</li>
<li>python39Packages.keystoneauth1</li>
<li>python39Packages.keystoneclient</li>
<li>python39Packages.novaclient</li>
<li>python39Packages.openstacksdk</li>
<li>python39Packages.os-service-types</li>
<li>python39Packages.osc-lib</li>
<li>python39Packages.oslo-config</li>
<li>python39Packages.oslo-context</li>
<li>python39Packages.oslo-i18n</li>
<li>python39Packages.oslo-log</li>
<li>python39Packages.oslo-serialization</li>
<li>python39Packages.oslo-utils</li>
<li>python39Packages.oslotest</li>
<li>stestr (python39Packages.stestr)</li>
</ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 134282` at 4acd0489 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/05a71294452c9ae00591aacaf2e704c3">openstackclient</a></li>
    <li>python38Packages.openstacksdk</li>
    <li><a href="https://gist.github.com/2f47b8967d52424f8179522ed89974a4">python38Packages.osc-lib</a></li>
    <li>python39Packages.openstacksdk</li>
    <li><a href="https://gist.github.com/ea24069d428c5565f682d13d5c480cd5">python39Packages.osc-lib</a></li>
  </ul>
</details>
<details>
  <summary>30 packages built successfully:</summary>
  <ul>
    <li>python38Packages.cinderclient</li>
    <li>python38Packages.debtcollector</li>
    <li>python38Packages.hacking</li>
    <li>python38Packages.keystoneauth1</li>
    <li>python38Packages.keystoneclient</li>
    <li>python38Packages.novaclient</li>
    <li>python38Packages.os-service-types</li>
    <li>python38Packages.oslo-config</li>
    <li>python38Packages.oslo-context</li>
    <li>python38Packages.oslo-i18n</li>
    <li>python38Packages.oslo-log</li>
    <li>python38Packages.oslo-serialization</li>
    <li>python38Packages.oslo-utils</li>
    <li>python38Packages.oslotest</li>
    <li>python38Packages.stestr</li>
    <li>python39Packages.cinderclient</li>
    <li>python39Packages.debtcollector</li>
    <li>python39Packages.hacking</li>
    <li>python39Packages.keystoneauth1</li>
    <li>python39Packages.keystoneclient</li>
    <li>python39Packages.novaclient</li>
    <li>python39Packages.os-service-types</li>
    <li>python39Packages.oslo-config</li>
    <li>python39Packages.oslo-context</li>
    <li>python39Packages.oslo-i18n</li>
    <li>python39Packages.oslo-log</li>
    <li>python39Packages.oslo-serialization</li>
    <li>python39Packages.oslo-utils</li>
    <li>python39Packages.oslotest</li>
    <li>stestr (python39Packages.stestr)</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `testscenarios`.
Near pkgs/development/python-modules/oslo-context/default.nix:3:21:
```
  |
3 | , stestr, oslotest, testscenarios }:
  |                     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `testscenarios`.
Near pkgs/development/python-modules/oslo-log/default.nix:3:21:
```
  |
3 | , stestr, oslotest, testscenarios }:
  |                     ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by angustrau** [comment]
Had to disable checks on `openstacksdk` as I couldn't get a small handful of tests to run consistently. I also renamed some packages to match their pypi names.

**Comment by SuperSandro2000** [comment]
> Had to disable checks on openstacksdk

``doCheck = false``?

**Comment by angustrau** [comment]
@SuperSandro2000 yep. I left a comment there. Hopefully upstream improves on the reliability at some point and we can turn it back on.

File: pkgs/development/python-modules/os-service-types/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    sha256 = "0v4chwr5jykkvkv4w7iaaic7gb06j6ziw7xrjlwkcf92m2ch501i";
+  };
+
+  postPatch = ''
+    # This test has circular dependency on keystoneauth1
+    rm os_service_types/tests/test_remote.py
+  '';
+
+  propagatedBuildInputs = [
+    pbr
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/os-service-types/default.nix:20]: Like all the other pbr's it most likely goes into nativeBuildInputs.

File: pkgs/development/python-modules/oslo-log/default.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+, pbr, oslo-config, oslo-context, oslo-i18n, oslo-utils, oslo-serialization, debtcollector, pyinotify, python-dateutil
+, stestr, oslotest }:
+
+buildPythonPackage rec {
+  pname = "oslo-log";
+  version = "4.6.0";
+
+  src = fetchPypi {
+    inherit version;
+    pname = "oslo.log";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/oslo-log/default.nix:11]:
```suggestion
    pname = "oslo.log";
    inherit version;
```

File: pkgs/development/python-modules/oslo-utils/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+, pbr, iso8601, oslo-i18n, pytz, netaddr, netifaces, debtcollector, pyparsing, packaging
+, stestr, oslotest, ddt, testscenarios }:
+
+buildPythonPackage rec {
+  pname = "oslo-utils";
+  version = "4.9.2";
+
+  src = fetchPypi {
+    inherit version;
+    pname = "oslo.utils";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/oslo-utils/default.nix:11]:
```suggestion
    pname = "oslo.utils";
    inherit version;
```

**Review by SuperSandro2000 — COMMENTED** [review]
I would suggest we create a maintainer group for all openstack things and then add you and myself to it.

**Comment by SuperSandro2000** [comment]
I also started a packaging effort to get openstack packaged. I maybe add my changes on top of yours so that we get this merged in the next days. You will keep your author ship and maintainer in it.

**Review by superherointj — COMMENTED** [review]
> I would suggest we create a maintainer group for all openstack things and then add you and myself to it.

Add me too, please. I have many parts of openstack packaged. I happen to have made duplicate and additional work here (that isn't merged yet). I am waiting this PR to land before making new PRs.

I have made a few different choices. Like, grouping openstack libraries in a single folder. OpenStack is ~400 python packages.

Would be nice if we had better coordination in this OpenStack project. @angustrau What other parts do you have ready?

Btw, thanks again Sandro for making yourself available. Really glad we have you here.

**Comment by angustrau** [comment]
Thanks @SuperSandro2000 @superherointj ! I'll try to make these changes this weekend. This is all I have packaged so far, but I have plans to complete all of the CLI plugins once merged.

**Comment by SuperSandro2000** [comment]
I did a separate PR for the team https://github.com/NixOS/nixpkgs/pull/137295

**Comment by SuperSandro2000** [comment]
I am going to open yet another PR which includes tests for all oslo components without infinite recursion. I hope that works for you two.

**Comment by SuperSandro2000** [comment]
Replaced by #134282 with tests enabled for everything.

I included everything except openstackclient-full. I think we should revisit that in the future depending on what else gets packaged. Maybe it is easier to just make a fake python package which propagates everything.

**Comment by SuperSandro2000** [comment]
I almost took everything from this PR except openstackclientFull. I am not sure how we want to design that and I would just see what works in the future for us.

**Pull request closed without merging**